pgxn/neon_walredo/walredoproc.c

```diff
@@ -574,11 +603,7 @@ ApplyRecord(StringInfo input_message)
 
 	/* note: the input must be aligned here */
 	record = (XLogRecord *) pq_getmsgbytes(input_message, sizeof(XLogRecord));
-
-	nleft = input_message->len - input_message->cursor;
-	if (record->xl_tot_len != sizeof(XLogRecord) + nleft)
-		elog(ERROR, "mismatch between record (%d) and message size (%d)",
-			 record->xl_tot_len, (int) sizeof(XLogRecord) + nleft);
+	input_message->cursor += record->xl_tot_len - sizeof(XLogRecord);
```

Comment:
advacing cursor here seems to be the right thing to do, but I don't understand why is it needed. Why doesn't the XLogBeginRead advance it?

Also, how will the line 575/604 work out to be aligned with full message transfers?

Reply:
> advacing cursor here seems to be the right thing to do, but I don't understand why is it needed. Why doesn't the XLogBeginRead advance it?

Because we receiver all wal-redo message with one pipe operation.  Original (stdio pipe) version receives each message individually, that it is why it doesn't need to advance cursor. It seems to be right optimization, especially taken in account fact that pageserver sends all message using one write operation.

> Also, how will the line 575/604 work out to be aligned with full message transfers?

Function `shmem_pipe_process_request` construct the whole message before passing it to walredo process.